tests/erc20_integration_test.go

```diff
@@ -5,70 +5,106 @@ package tests
 
 import (
 	"context"
-	"crypto/ecdsa"
 	"math/big"
 	"testing"
 
-	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	ethcommon "github.com/ethereum/go-ethereum/common"
 	ethcrypto "github.com/ethereum/go-ethereum/crypto"
-	"github.com/ethereum/go-ethereum/ethclient"
 	"github.com/stretchr/testify/require"
 
 	"github.com/athanorlabs/atomic-swap/common"
-	"github.com/athanorlabs/atomic-swap/common/types"
+	"github.com/athanorlabs/atomic-swap/common/rpctypes"
 	contracts "github.com/athanorlabs/atomic-swap/ethereum"
-	"github.com/athanorlabs/atomic-swap/ethereum/block"
+	"github.com/athanorlabs/atomic-swap/ethereum/extethclient"
+	"github.com/athanorlabs/atomic-swap/rpcclient"
 )
 
-func setupXMRTakerAuth(t *testing.T) (*bind.TransactOpts, *ethclient.Client, *ecdsa.PrivateKey) {
-	conn, chainID := NewEthClient(t)
-	pk, err := ethcrypto.HexToECDSA(common.DefaultPrivKeyXMRTaker)
+// deploys ERC20Mock.sol and assigns the whole token balance to the XMRTaker default address.
+func deployERC20Mock(t *testing.T) ethcommon.Address {
+	ctx := context.Background()
+	aliceKey, err := ethcrypto.HexToECDSA(common.DefaultPrivKeyXMRTaker)
 	require.NoError(t, err)
-	auth, err := bind.NewKeyedTransactorWithChainID(pk, chainID)
+
+	ec := extethclient.CreateTestClient(t, aliceKey)
+	txOpts, err := ec.TxOpts(ctx)
 	require.NoError(t, err)
-	return auth, conn, pk
-}
 
-// deploys ERC20Mock.sol and assigns the whole token balance to the XMRTaker default address.
-func deployERC20Mock(t *testing.T) ethcommon.Address {
-	auth, conn, pkA := setupXMRTakerAuth(t)
-	pub := pkA.Public().(*ecdsa.PublicKey)
-	addr := ethcrypto.PubkeyToAddress(*pub)
+	const (
+		initialTokenBalance = 1000 // standard units
+		decimals            = 18
+	)
+	tenToDecimals := new(big.Int).Exp(big.NewInt(10), big.NewInt(decimals), nil)
+	totalSupply := new(big.Int).Mul(big.NewInt(initialTokenBalance*2), tenToDecimals)
+	halfSupply := new(big.Int).Mul(big.NewInt(initialTokenBalance), tenToDecimals)
+
+	erc20Addr, erc20Tx, tokenContract, err := contracts.DeployERC20Mock(
+		txOpts,
+		ec.Raw(),
+		"ERC20Mock",
+		"MOCK",
+		decimals,
+		ec.Address(),
+		totalSupply,
+	)
+	require.NoError(t, err)
+	MineTransaction(t, ec.Raw(), erc20Tx)
 
-	decimals := new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)
-	balance := new(big.Int).Mul(big.NewInt(9999999), decimals)
-	erc20Addr, erc20Tx, _, err := contracts.DeployERC20Mock(auth, conn, "ERC20Mock", "MOCK", addr, balance)
+	// Query Charlie's Ethereum address
+	charlieCli := rpcclient.NewClient(ctx, defaultCharlieSwapdEndpoint)
+	balResp, err := charlieCli.Balances(nil)
 	require.NoError(t, err)
-	_, err = block.WaitForReceipt(context.Background(), conn, erc20Tx.Hash())
+	charlieAddr := balResp.EthAddress
+
+	// Transfer half of the supply to Charlie (using Alice's extended ethereum client)
```

Comment:
why does charlie need tokens? he doesn't swap right?

Reply:
There is a test where Alice and Charlie try to take Bob's offer at the same time.